vyper/semantics/types/module.py

```diff
@@ -559,3 +559,19 @@ def immutable_section_bytes(self):
     @cached_property
     def interface(self):
         return InterfaceT.from_ModuleT(self)
+
+    @cached_property
+    def is_stateless(self):
```

Comment:
i think this definition omits `reentrancy` locks - for cancun those are created inside `transient` which we define as state. the locks aren't implemented as variables so this check would miss them

to further support the statement above, consider following:
```vyper
#main.vy
import lib1

#initializes: lib1

exports: lib1.foo
```

```vyper
#lib1.vy
@external
@nonreentrant
def foo() -> uint256:
    return 5
```

which yields:
```
vyper.exceptions.ImmutableViolation: Cannot access `lib1` state!
  note that use of the `@nonreentrant` decorator is also considered state access

  contract "tests/custom/test.vy:5", line 5:9 
       4
  ---> 5 exports: lib1.foo
  ----------------^
       6

  (hint: add `uses: lib1` or `initializes: lib1` as a top-level statement to your contract)
```